src/registry.rs

```diff
@@ -0,0 +1,274 @@
+    ///
+    /// For example
+    ///
+    /// ```
+    /// # use std::sync::Arc;
+    /// # use url::Url;
+    /// # use object_store::memory::InMemory;
+    /// # use object_store::ObjectStore;
+    /// # use object_store::prefix::PrefixStore;
+    /// # use object_store::registry::{DefaultObjectStoreRegistry, ObjectStoreRegistry};
+    /// #
+    /// let registry = DefaultObjectStoreRegistry::new();
+    ///
+    /// let bucket1 = Arc::new(InMemory::new()) as Arc<dyn ObjectStore>;
+    /// let base = Url::parse("s3://bucket1/").unwrap();
+    /// registry.register(base, bucket1.clone());
+    ///
+    /// let url = Url::parse("s3://bucket1/path/to/object").unwrap();
+    /// let (ret, path) = registry.resolve(&url).unwrap();
+    /// assert_eq!(path.as_ref(), "path/to/object");
+    /// assert!(Arc::ptr_eq(&ret, &bucket1));
+    ///
+    /// let bucket2 = Arc::new(InMemory::new()) as Arc<dyn ObjectStore>;
+    /// let base = Url::parse("https://s3.region.amazonaws.com/bucket").unwrap();
+    /// registry.register(base, bucket2.clone());
+    ///
+    /// let url = Url::parse("https://s3.region.amazonaws.com/bucket/path/to/object").unwrap();
+    /// let (ret, path) = registry.resolve(&url).unwrap();
+    /// assert_eq!(path.as_ref(), "path/to/object");
+    /// assert!(Arc::ptr_eq(&ret, &bucket2));
+    ///
+    /// let bucket3 = Arc::new(PrefixStore::new(InMemory::new(), "path")) as Arc<dyn ObjectStore>;
+    /// let base = Url::parse("https://s3.region.amazonaws.com/bucket/path").unwrap();
+    /// registry.register(base, bucket3.clone());
+    ///
+    /// let url = Url::parse("https://s3.region.amazonaws.com/bucket/path/to/object").unwrap();
+    /// let (ret, path) = registry.resolve(&url).unwrap();
+    /// assert_eq!(path.as_ref(), "to/object");
+    /// assert!(Arc::ptr_eq(&ret, &bucket3));
+    /// ```
+    fn resolve(&self, url: &Url) -> crate::Result<(Arc<dyn ObjectStore>, Path)>;
+}
+
+/// Error type for [`DefaultObjectStoreRegistry`]
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+enum Error {
+    #[error("ObjectStore not found")]
+    NotFound,
+
+    #[error("Error parsing URL path segment")]
+    InvalidPart(#[from] InvalidPart),
+}
+
+impl From<Error> for crate::Error {
+    fn from(value: Error) -> Self {
+        Self::Generic {
+            store: "ObjectStoreRegistry",
+            source: Box::new(value),
+        }
+    }
+}
+
+/// An [`ObjectStoreRegistry`] that uses [`parse_url_opts`] to create stores based on the environment
+#[derive(Debug, Default)]
+pub struct DefaultObjectStoreRegistry {
+    map: RwLock<HashMap<String, PathEntry>>,
+}
+
+#[derive(Debug, Default)]
+struct PathEntry {
+    store: Option<Arc<dyn ObjectStore>>,
+    children: HashMap<String, Self>,
+}
+
+impl PathEntry {
+    /// Lookup a store based on URL path
+    ///
+    /// Returns the store and its path segment depth
+    fn lookup(&self, to_resolve: &Url) -> Option<(&Arc<dyn ObjectStore>, usize)> {
```

Comment:
This does mean that if people register really long paths, this process will be rather inefficient, however, in most cases I think this should be acceptable.